python/instrumentation/openinference-instrumentation-openai/src/openinference/instrumentation/openai/_request.py

```diff
@@ -0,0 +1,358 @@
+import logging
+from abc import ABC
+from contextlib import contextmanager
+from types import MappingProxyType
+from typing import (
+    Any,
+    Awaitable,
+    Callable,
+    Dict,
+    Hashable,
+    Iterator,
+    Mapping,
+    Optional,
+    Tuple,
+)
+
+from openinference.instrumentation.openai._extra_attributes_from_request import (
+    _get_extra_attributes_from_request,
+)
+from openinference.instrumentation.openai._extra_attributes_from_response import (
+    _get_extra_attributes_from_response,
+)
+from openinference.instrumentation.openai._response_accumulator import (
+    _ChatCompletionAccumulator,
+    _CompletionAccumulator,
+)
+from openinference.instrumentation.openai._stream import (
+    _ResponseAccumulator,
+    _Stream,
+)
+from openinference.instrumentation.openai._utils import (
+    _as_input_attributes,
+    _as_output_attributes,
+    _finish_tracing,
+    _io_value_and_type,
+)
+from openinference.instrumentation.openai._with_span import _WithSpan
+from openinference.semconv.trace import SpanAttributes
+from opentelemetry import context as context_api
+from opentelemetry import trace as trace_api
+from opentelemetry.context import _SUPPRESS_INSTRUMENTATION_KEY
+from opentelemetry.trace import INVALID_SPAN
+from opentelemetry.util.types import AttributeValue
+
+from openai import AsyncStream, Stream
+from openai.types import Completion, CreateEmbeddingResponse
+from openai.types.chat import ChatCompletion
+
+__all__ = (
+    "_Request",
+    "_AsyncRequest",
+)
+
+logger = logging.getLogger(__name__)
+logger.addHandler(logging.NullHandler())
+
+_LLM_SPAN_KIND = "LLM"
+_EMBEDDING_SPAN_KIND = "EMBEDDING"
+
+
+class _WithTracer(ABC):
+    __slots__ = (
+        "_tracer",
+        "_include_extra_attributes",
+    )
+
+    def __init__(
+        self,
+        tracer: trace_api.Tracer,
+        include_extra_attributes: bool = True,
+    ) -> None:
+        self._tracer = tracer
+        self._include_extra_attributes = include_extra_attributes
+
+    @contextmanager
+    def _start_as_current_span(
+        self,
+        span_name: str,
+        cast_to: type,
+        request_options: Mapping[str, Any],
+    ) -> Iterator[_WithSpan]:
+        span_kind = _EMBEDDING_SPAN_KIND if cast_to is CreateEmbeddingResponse else _LLM_SPAN_KIND
```

Comment:
yea, agree. i'll update. On the other hand maybe the span name should come from a higher level function: by the time the Request function is called, we don't really know who called it and for what purpose. An alternative is to also monkey patch those higher level callers, but instead of tracing them per se, just have them put the real `span_name` into the `context` and pass it down to `request` (and then use the fallback in request if `span_name` doesn't exist.